IdentityCore/src/throttling/metadata/MSIDThrottlingMetaDataCache.m

```diff
@@ -0,0 +1,100 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDThrottlingMetaDataCache.h"
+#import "MSIDCacheKey.h"
+#import "MSIDRequestContext.h"
+#import "MSIDThrottlingMetaData.h"
+#import "MSIDExtendedTokenCacheDataSource.h"
+#import "MSIDConstants.h"
+#import "MSIDCacheItemJsonSerializer.h"
+#import "MSIDKeychainTokenCache.h"
+#import "NSDate+MSIDExtensions.h"
+
+@implementation MSIDThrottlingMetaDataCache
+
++ (MSIDThrottlingMetaData *)getThrottlingMetadataWithAccessGroup:(NSString *)accessGroup
+                                                         Context:(id<MSIDRequestContext>)context
+                                                           error:(NSError*__nullable*__nullable)error
+{
+    MSIDThrottlingMetaData *result = nil;
+    id<MSIDExtendedTokenCacheDataSource> datasource = [MSIDThrottlingMetaDataCache dataSource:accessGroup error:error];
+    if (!datasource) return nil;
+    NSArray *jsonObjects = [datasource jsonObjectsWithKey:[MSIDThrottlingMetaDataCache throttlingMetadataCacheKey]
+                                               serializer:[MSIDCacheItemJsonSerializer new]
+                                                  context:context error:error];
+    if (![jsonObjects count])
+    {
+        return nil;
+    }
+    result = [[MSIDThrottlingMetaData alloc] initWithJSONDictionary:[jsonObjects[0] jsonDictionary] error:error];
+    return result;
+}
+
++ (BOOL)updateLastRefreshTimeWithAccessGroup:(NSString *)accessGroup
+                                     Context:(id<MSIDRequestContext>)context
+                                       error:(NSError*__nullable*__nullable)error
+{
+    MSIDThrottlingMetaData *metadata = [[MSIDThrottlingMetaData alloc] init];
+    metadata.lastRefreshTime = [[NSDate new] msidDateToTimestamp];
+    MSIDJsonObject *metadataJSONObj = [[MSIDJsonObject alloc] initWithJSONDictionary:metadata.jsonDictionary error:error];
+    if (!metadataJSONObj)
+    {
+        return NO;
+    }
+    
+    id<MSIDExtendedTokenCacheDataSource> datasource = [MSIDThrottlingMetaDataCache dataSource:accessGroup error:error];
+    if (!datasource)
+    {
+        return NO;
+    }
+
+    return [datasource saveJsonObject:metadataJSONObj
+                           serializer:[MSIDCacheItemJsonSerializer new]
+                                  key:[MSIDThrottlingMetaDataCache throttlingMetadataCacheKey]
+                              context:context
+                                error:error];
+}
+
++ (id<MSIDExtendedTokenCacheDataSource>)dataSource:(NSString *)accessGroup error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    return [[MSIDKeychainTokenCache alloc] initWithGroup:accessGroup error:error];
```

Comment:
Same comment - let's try to not create new one every time. Possible to reuse token cache object from the request object (it should already have one).